### PR TITLE
adding a sample for custom recoverability

### PR DESCRIPTION
**Pull request by soujay**
This PR adds a sample on how to implement custom recoverability policy

### CODE REVIEW

File: samples/custom-recoverability/Core_7/Server/MyHandler.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+public class MyHandler :
+    IHandleMessages<MyMessage>
+{
+    static ILog log = LogManager.GetLogger<MyHandler>();
+
+    public Task Handle(MyMessage message, IMessageHandlerContext context)
+    {
+        log.Info($"Message received. Id: {message.Id}");
+       // throw new ArgumentNullException("Uh oh - something went wrong....");
+        // throw new DivideByZeroException("DivideByZeroException - something went wrong....");
```
Comment (WilliamBZA) [inline samples/custom-recoverability/Core_7/Server/MyHandler.cs:16]: Indentation is weird here. Why are these commented out?

File: samples/custom-recoverability/Core_7/Server/Program.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        var endpointConfiguration = new EndpointConfiguration("Samples.CustomRecoverability.Server");
+        endpointConfiguration.UsePersistence<LearningPersistence>();
+        endpointConfiguration.UseTransport<LearningTransport>();
+        var recoverability = endpointConfiguration.Recoverability();
+
+        #region disable
+        //recoverability.Delayed(settings =>
+        //{
+        //    settings.NumberOfRetries(0);
+        //});
```
Comment (WilliamBZA) [inline samples/custom-recoverability/Core_7/Server/Program.cs:20]: Why is this commented out?

File: samples/custom-recoverability/Core_8/Server/MyHandler.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+public class MyHandler :
+    IHandleMessages<MyMessage>
+{
+    static ILog log = LogManager.GetLogger<MyHandler>();
+
+    public Task Handle(MyMessage message, IMessageHandlerContext context)
+    {
+        log.Info($"Message received. Id: {message.Id}");
+        // throw new ArgumentNullException("Uh oh - something went wrong....");
+       // throw new DivideByZeroException("DivideByZeroException - something went wrong....");
```
Comment (WilliamBZA) [inline samples/custom-recoverability/Core_8/Server/MyHandler.cs:16]: Indentation is weird here. Why is this commented out?

File: samples/custom-recoverability/Core_8/Server/Program.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        var endpointConfiguration = new EndpointConfiguration("Samples.CustomRecoverability.Server");
+        endpointConfiguration.UsePersistence<LearningPersistence>();
+        endpointConfiguration.UseTransport(new LearningTransport());
+        var recoverability = endpointConfiguration.Recoverability();
+
+        #region disable
+        //recoverability.Delayed(settings =>
+        //{
+        //    settings.NumberOfRetries(0);
+        //});
```
Comment (WilliamBZA) [inline samples/custom-recoverability/Core_8/Server/Program.cs:20]: Why is this commented out?

File: samples/custom-recoverability/Core_7/Server/MyHandler.cs (hunk 1 above)
Comment (soujay) [reply samples/custom-recoverability/Core_7/Server/MyHandler.cs]: These are commented out to begin with and the sample doc has instructions when to uncomment them and what happens when its done.

File: samples/custom-recoverability/Core_7/Server/Program.cs (hunk 2 above)
Comment (soujay) [reply samples/custom-recoverability/Core_7/Server/Program.cs:20]: These are commented out to begin with and the sample doc has instructions when to uncomment them and what happens when its done.

File: samples/custom-recoverability/Core_8/Server/MyHandler.cs (hunk 3 above)
Comment (soujay) [reply samples/custom-recoverability/Core_8/Server/MyHandler.cs]: These are commented out to begin with and the sample doc has instructions when to uncomment them and what happens when its done.

File: samples/custom-recoverability/Core_8/Server/Program.cs (hunk 4 above)
Comment (soujay) [reply samples/custom-recoverability/Core_8/Server/Program.cs:20]: These are commented out to begin with and the sample doc has instructions when to uncomment them and what happens when its done.

File: samples/custom-recoverability/sample.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+- nservicebus/nservicebus-step-by-step-guide-custom-recoverability
+related:
+- nservicebus/recoverability
+- nservicebus/recoverability/custom-recoverability-policy
+- samples/errorhandling
+- samples/faulttolerance
+
+---
+
+This sample demonstrates how to use custom recoverability policy. Here the default recoverability is invoked when a specifc exception type happens and at all other times it uses a custom recoverability action.
```
Comment (WilliamBZA) [inline samples/custom-recoverability/sample.md:16]:
```suggestion
This sample demonstrates how to use custom recoverability policy. Here the default recoverability is invoked when a specific exception type happens and at all other times it uses a custom recoverability action.
```

File: samples/custom-recoverability/sample.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+**Server output**
+
+```
+Press any key to exit
+Message received. Id: 87283b82a892456d81eb1bfa05fb72e2
+```
+
+## Fault tolerance with custom recoverability
+
+In the 'Server' application, open `Program.cs`. There is a custom policy "MyCustomRetryPolicy" where the default NServivceBus [recoverability policy](/nservicebus/recoverability/) is invoked when an "ArgumentNullException" exception is encountered and at all other times the message is not retried and sent directly to the "error" queue.
```
Comment (WilliamBZA) [inline samples/custom-recoverability/sample.md:46]:
```suggestion
In the 'Server' application, open `Program.cs`. There is a custom policy `MyCustomRetryPolicy` where the default NServivceBus [recoverability policy](/nservicebus/recoverability/) is invoked when an `ArgumentNullException` exception is encountered and at all other times the message is not retried and sent directly to the "error" queue.
```

File: samples/custom-recoverability/sample.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+snippet: addcustomheaders
+
+### Make the handler fail
+
+In the 'Server' application, open `MyHandler.cs`. Uncomment the `throw new ArgumentNullException` line.
+
+snippet: MyHandler
+
+As per the custom recoverability policy, when the solution is run, the exception is thrown and the default [recoverability policy](/nservicebus/recoverability/) will be invoked before finalling moving to the "error" queue.
```
Comment (WilliamBZA) [inline samples/custom-recoverability/sample.md:64]:
```suggestion
As per the custom recoverability policy, when the solution is run, the exception is thrown and the default [recoverability policy](/nservicebus/recoverability/) will be invoked before finally moving to the "error" queue.
```

File: samples/custom-recoverability/sample.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+INFO  Message received. Id: 76d8b5c3-c41b-4179-b995-4e68f3c5b7eb
+ERROR Moving message '1e0d3ee9-3ab0-4be9-bf45-b0330060b9e8' to the error queue 'error' because processing failed due to an exception:
+System.ArgumentNullException: Value cannot be null. (Parameter 'Uh oh - something went wrong....')
+```
+
+Delayed Retries can be turned off by uncommenting the below line in the Server Program.cs:
+
+snippet: disable
+
+When the sample is re-run the sample, after successive retries, the message is sent to the error queue without the delayed retries.
```
Comment (WilliamBZA) [inline samples/custom-recoverability/sample.md:92]:
```suggestion
Now when the sample is re-run, the message is sent to the error queue without the delayed retries after successive immediate retries.
```

Is this correct? I struggled to parse what this sentence is trying to say.

**Review by WilliamBZA — CHANGES_REQUESTED** [review]

File: samples/custom-recoverability/sample.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+**Server output**
+
+```
+Press any key to exit
+Message received. Id: 87283b82a892456d81eb1bfa05fb72e2
+```
+
+## Fault tolerance with custom recoverability
+
+In the 'Server' application, open `Program.cs`. There is a custom policy `MyCustomRetryPolicy` where the default NServivceBus [recoverability policy](/nservicebus/recoverability/) is invoked when an `ArgumentNullException` exception is encountered and at all other times the message is not retried and sent directly to the "error" queue.
```
Comment (WilliamBZA) [inline samples/custom-recoverability/sample.md:46]:
```suggestion
In the 'Server' application, open `Program.cs`. There is a custom policy `MyCustomRetryPolicy` where the default NServiceBus [recoverability policy](/nservicebus/recoverability/) is invoked when an `ArgumentNullException` exception is encountered and at all other times the message is not retried and sent directly to the "error" queue.
```

File: samples/custom-recoverability/sample.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+System.ArgumentNullException: Value cannot be null. (Parameter 'Uh oh - something went wrong....')
+```
+
+Delayed Retries can be turned off by uncommenting the below line in the Server Program.cs:
+
+snippet: disable
+
+Now when the sample is re-run, the message is sent to the error queue without the delayed retries after successive immediate retries.
+
+Now, in the 'Server' application, open `MyHandler.cs`. Comment out `throw new ArgumentNullException` and uncomment the ` throw new DivideByZeroException` line. When the solution is run, and the exception is thrown  the message is directly moved to the "error" queue without any retries ( as per the custom policy).
```
Comment (WilliamBZA) [inline samples/custom-recoverability/sample.md:94]:
```suggestion
Now, in the "Server" application, open `MyHandler.cs`. Comment out `throw new ArgumentNullException` and uncomment the `throw new DivideByZeroException` line. When the solution is run and the exception is thrown, the message is directly moved to the "error" queue without any retries (as per the custom policy).
```

**Review by WilliamBZA — CHANGES_REQUESTED** [review]

**Review by WilliamBZA — APPROVED** [review]

**Pull request merged**